HTTPCLIENT-2277: Improve Age Header Handling and Calculation in Accordance with RFC9111.

arturobernalg opened the pull request:
This commit enhances the processing and calculation of the "Age" response header field to comply fully with Sections 5.1 
and 4.2.3 of RFC9111.

Changes include:

- Parsing of the Age field value as a non-negative integer, representing time in seconds. 
- Handling of messages with list-based Age field values, using the first member and discarding subsequent ones.
- Ignoring of invalid field values (e.g., containing something other than a non-negative integer).
- The presence of an Age header field implies that the response was not generated or validated by the origin server for this request.
- Improved/Fixing the age calculation algorithm considering the apparent age, response delay, corrected age value, and current age to align it with the standard.
-

ok2c commented on httpclient5-cache/src/test/java/org/apache/hc/client5/http/impl/cache/TestProtocolRequirements.java:
```diff
@@ -2123,7 +2123,13 @@ public void testAgeHeaderPopulatedFromCacheEntryCurrentAge() throws Exception {
+        // We then add the "residentTime" which is "now - responseTime",
+        // which is the current time minus the time the cache entry was created. In this case, that is 8 seconds.
+        // So, the total age is "corrected_initial_age" + "residentTime" = 2 + 8 = 10 seconds.
+        Assertions.assertEquals("10", result.getFirstHeader("Age").getValue());
```
@arturobernalg  I find `ContainsHeaderMatcher` look a bit nicer here .

arturobernalg replied:
@ok2c 
The current implementation of ContainsHeaderMatcher does not directly support the object types we're working with here, such as ClassicHttpResponse. Adapting it to these types may require significant changes to the existing code. Maybe I'm missing something.

ok2c replied:
@arturobernalg Ah, i fixed the matcher in my branch but not in 5.3.x. Please pull the latest changes.

arturobernalg replied:
@ok2c  Change applied.

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheValidityPolicy.java:
```diff
@@ -206,28 +206,51 @@ protected TimeValue getApparentAge(final HttpCacheEntry entry) {
+                    }
+                    result = result * 10 + digit;
+                    if (result > Integer.MAX_VALUE) {
+                        if (LOG.isWarnEnabled()) {
```
@arturobernalg Please reduce priority of those logs to DEBUG. There is nothing more annoying than a noisy low level transport library in a huge application.

arturobernalg replied:
changed.

ok2c commented:
@arturobernalg The way I read `RFC 9111 5.1` the client would be able to handle `Age` values like `123, 123, 0` or `123, 123, crap` and take the first value from the token list as long as it is valid.

ok2c reviewed (changes requested):
@arturobernalg What happens when you get `Age: 123crap` header? 

It is great that you are trying to make your code efficient but in this case it is getting a bit too complex. Please use the support methods from `CacheSupport` I just added, parser the header into a sequence of tokens, take the first token and try to convert into an int.

arturobernalg commented:
> @arturobernalg What happens when you get `Age: 123crap` header?
> 
> It is great that you are trying to make your code efficient but in this case it is getting a bit too complex. Please use the support methods from `CacheSupport` I just added, parser the header into a sequence of tokens, take the first token and try to convert into an int.



> @arturobernalg What happens when you get `Age: 123crap` header?
> 
> It is great that you are trying to make your code efficient but in this case it is getting a bit too complex. Please use the support methods from `CacheSupport` I just added, parser the header into a sequence of tokens, take the first token and try to convert into an int.

wait. I dint see this comment./.

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheValidityPolicy.java:
```diff
@@ -206,28 +207,60 @@ protected TimeValue getApparentAge(final HttpCacheEntry entry) {
-                hdrAge = Long.parseLong(hdr.getValue());
-                if (hdrAge < 0) {
-                    hdrAge = MAX_AGE.toSeconds();
+        final Header[] ages = entry.getHeaders(HttpHeaders.AGE);
```
RFC 9111, 5.1
```
Although it is defined as a singleton header field, a cache encountering a message with a list-based
Age field value SHOULD use the first member of the field value, discarding subsequent ones.
If the field value (after discarding additional members, as per above) is invalid (e.g., it contains
something other than a non-negative integer), a cache SHOULD ignore the field.
```

@arturobernalg This looks wrong. We are supposed to take the first value, not the last one, aren't we? Get the first header and get the first token out of it. That is it.

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheValidityPolicy.java:
```diff
@@ -206,28 +207,62 @@ protected TimeValue getApparentAge(final HttpCacheEntry entry) {
-                hdrAge = Long.parseLong(hdr.getValue());
-                if (hdrAge < 0) {
-                    hdrAge = MAX_AGE.toSeconds();
+                final AtomicLong ageValue = new AtomicLong(0);
```
@arturobernalg You are still parsing all tokens instead of one (which is wasteful) and taking the last value from the sequence instead of the first. Please add test cases for various malformed values and make sure they all pass.

It is perfectly fine to use Long#parseLong. It is not going to be any less efficient.

arturobernalg replied:
@ok2c 
Should be fine now.

ok2c replied:
@arturobernalg Please use this code to extract the first token. It is massively simpler and is likely more efficient.
```
            final AtomicReference<String> firstToken = new AtomicReference<>();
            CacheSupport.parseTokens(age, token -> {
                firstToken.compareAndSet(null, token);
            });
            final String s = firstToken.get();
```
If you want to handle delta-seconds `"1" + Long.MAX_VALUE` please do it consistently for all headers, not just `Age`. For now I think accepting only valid long values that can be parsed with `Long#parseLong` should be good enough.

arturobernalg replied:
@ok2c 
Changed

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheValidityPolicy.java:
```diff
@@ -206,28 +207,53 @@ protected TimeValue getApparentAge(final HttpCacheEntry entry) {
+                    return ageValue;
+                }
+            } catch (final NumberFormatException ex) {
+                if (LOG.isWarnEnabled()) {
```
@arturobernalg Looks very good now. Please do reduce this one to DEBUG priority.
